backport: Use FastRandomContext for all tests

xanimo opened the pull request:
This is a follow up to #3487 ~~and #3497 which are requirements~~ which is a requirement needed in order to get #3229 in.

xanimo commented:
(I will rebase accordingly as to give attribution to chromatic and ryanofsky as the review process continues.)

xanimo commented:
Force pushed to give attribution to ryanofsky. Just noticing an extraneous `#include <boost/filesystem.hpp>` post #3488 so going to force push again shortly.

xanimo commented:
Force pushed to remove [extraneous boost headers](https://github.com/dogecoin/dogecoin/pull/3511/commits/f92324ca4c7da07d84bcde45bf6e11b5d1ece17a#diff-b6871acd6daf76c8a8a07cb8d0c472c1b4d3ebadb4115118f62aae3f3c0bd5a1L31-L32)

patricklodder reviewed (commented):
Reviewing commit by commit. 

I think that the "squashme" db5041005a should be squashed.

-------
Progress:
- [x] dc48e55de0 Introduce FastRandomContext::randbool()
- [x] caba928bbc FastRandom benchmark
- [x] 1fd572e0c7 Add ChaCha20
- [x] 85aae7de3d util: Specific GetOSRandom for Linux/FreeBSD/OpenBSD
- [x] db5041005a squashme: comment that NUM_OS_RANDOM_BYTES should not be changed lightly
- [x] aef09477f6 sanity: Move OS random to sanity check function
- [ ] 1fad144620 random: Add fallback if getrandom syscall not available
- [ ] e40241115e Switch FastRandomContext to ChaCha20
- [ ] ca9200c141 Add a FastRandomContext::randrange and use it
- [ ] f92324ca4c Make test_bitcoin.cpp compatible with Qt Test framework
- [ ] 7ee01e0f94 Add FastRandomContext::rand256() and ::randbytes()
- [ ] 03287224f7 Merge test_random.h into test_bitcoin.h
- [ ] 16758ea1cb Add various insecure_rand wrappers for tests
- [ ] 37c30e757d scripted-diff: use insecure_rand256/randrange more
- [ ] b4236f7806 Replace more rand() % NUM by randranges
- [ ] 50f01c72cc Replace rand() & ((1 << N) - 1) with randbits(N)
- [ ] 70095888d9 Use randbits instead of ad-hoc emulation in prevector tests
- [ ] af576f99f5 scripted-diff: Use randbits/bool instead of randrange where possible

xanimo commented:
Force pushed to add squashed commit db5041005aa84cbd97568800a6501bccd6d8063b to 28b9d89a50fa6a137c9558f90126dbf24dba6b11 and cherry-picks of https://github.com/dogecoin/dogecoin/commit/ee2d10ad0c0e04d0b9da4535a6fff265ac2501e5 and https://github.com/bitcoin/bitcoin/commit/c13c97db. I tried to add attribution to chromatic for their cherry-picks but it seems it didn't work.

patricklodder commented on src/random.cpp:
```diff
@@ -164,6 +167,15 @@ void GetOSRand(unsigned char *ent32)
     }
+#elif defined(HAVE_GETENTROPY_RAND) && defined(MAC_OSX)
+    // We need a fallback for OSX < 10.12
+    if (&getentropy != NULL) {
```
`nullptr`

patricklodder commented on configure.ac:
```diff
@@ -648,15 +648,8 @@ AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <unistd.h>
-)
-
-AC_MSG_CHECKING(for getentropy via random.h)
-AC_COMPILE_IFELSE([AC_LANG_PROGRAM([[#include <unistd.h>
```
I think this deletes a check too many? Not documented in commit message.

patricklodder commented on src/random.cpp:
```diff
@@ -159,14 +156,16 @@ void GetOSRand(unsigned char *ent32)
-    }
+    arc4random_buf(ent32, NUM_OS_RANDOM_BYTES);
+    // Silence a compiler warning about unused function.
+    (void)GetDevURandom;
```
This is not documented in the commit message. Basically replaces the entire OpenBSD implementation.

patricklodder reviewed (commented):
- 7977dd735d: why is this commit in this PR / why is this change relevant for the preceding / succeeding commits?
- 1552823cf3: this scripted diff fails for me (GNU sed 4.8). Either remove the script and the "scripted-diff:" annotation from the commit
, or make it so that the script works.
- nit: you've missed some pick references on the first couple commits (would be appreciated to have those, so that I can compare the picks with the originals.

-----

Post-rebase commit checklist:

- [x] 1:  dc48e55de0 =  1:  95cd06175b Introduce FastRandomContext::randbool()
- [x] 2:  caba928bbc =  2:  ed5ff1b27f FastRandom benchmark
- [x] 3:  1fd572e0c7 =  3:  e9a37818fa Add ChaCha20
- [x] 4:  85aae7de3d !  4:  28b9d89a50 util: Specific GetOSRandom for Linux/FreeBSD/OpenBSD
- [x] 5:  db5041005a <  -:  ---------- squashme: comment that NUM_OS_RANDOM_BYTES should not be changed lightly
- [x] 6:  aef09477f6 =  5:  814a15a4b1 sanity: Move OS random to sanity check function
- [x] 7:  1fad144620 =  6:  1cf8374af0 random: Add fallback if getrandom syscall not available
- [x] 8:  e40241115e =  7:  2ad516b8ec Switch FastRandomContext to ChaCha20
- [x] 9:  ca9200c141 =  8:  9b8d168293 Add a FastRandomContext::randrange and use it
- [x] 10:  f92324ca4c =  9:  7977dd735d Make test_bitcoin.cpp compatible with Qt Test framework
- [x] 11:  7ee01e0f94 = 10:  18e7aa60e0 Add FastRandomContext::rand256() and ::randbytes()
- [x] 12:  03287224f7 = 11:  6d749d1b97 Merge test_random.h into test_bitcoin.h
- [x] 13:  16758ea1cb = 12:  1c00a417d5 Add various insecure_rand wrappers for tests
- [x] 14:  37c30e757d = 13:  5fa8490a44 scripted-diff: use insecure_rand256/randrange more
- [x] 15:  b4236f7806 = 14:  64a1b72a19 Replace more rand() % NUM by randranges
- [x] 16:  50f01c72cc = 15:  e319e21965 Replace rand() & ((1 << N) - 1) with randbits(N)
- [x] 17:  70095888d9 = 16:  d80429400b Use randbits instead of ad-hoc emulation in prevector tests
- [ ] 18:  af576f99f5 = 17:  1552823cf3 scripted-diff: Use randbits/bool instead of randrange where possible
- [x] -:  ---------- > 18:  bd83403730 Check if sys/random.h is required for getentropy on OSX.
- [x] -:  ---------- > 19:  1873fb83e0 random: getentropy on macOS does not need unistd.h

xanimo commented:
Force pushed reworked branch. I think I addressed all of the above but let me know if I missed anything.

patricklodder commented:
I've tested performance for all tests that were edited between this PR at f9cd19f6b8 and 1.15.0-dev at 497c6026fc (except random tests as that's new) on an M1 with Apple clang 14.

script used icw `hyperfine`:
```bash
for t in DoS_tests addrman_tests blockencodings_tests bloom_tests checkqueue_tests coins_tests \
            crypto_tests cuckoocache_tests dbwrapper_tests merkle_tests miner_tests pmt_tests \
            pow_tests PrevectorTests script_tests sighash_tests skiplist_tests util_tests versionbits_tests; do 
  echo "'src/test/test_dogecoin -t $t'"; 
done | xargs -n1 hyperfine --warmup 2 -r 100 --style none --export-csv /dev/stdout --
```


This is consistent with what I measured with bench earlier and potentially can get a speedup from an unroll later, as discussed in #3487.

xanimo commented:
Still waking up but I think I've addressed your suggestions. Might need to reword the commit message more though. Let me know if what I wrote meets your approval.

patricklodder reviewed (approved):
tested ACK - tested on arm64 macOS Sonoma and x86_64 Ubuntu Jammy

Note: scope is greater than just tests (and bench), this also affects randomness calls when:
1. Selecting random peers from our peer database (`CAddrman::Select_()`)
2. Replacing a random entry in a maxed out list of peers to send out  (`CNode::PushAddress()`)
3. Random order of selecting available utxos in wallet (`ApproximateBestSubset()` in `wallet.cpp`)
